### ISSUE TITLE
com.flutter_app.MainActivity cannot be cast to io.flutter.app.FlutterActivity After flutter upgrade v1.12.13+hotfix.5

### ISSUE THREAD

**Issue by AmolGangadhare** [flutter/flutter#47127]
In my plugin, I need to cast activity to FlutterActivity. But after upgrading the flutter getting this error.

## Steps to Reproduce

1. Run app

**Target Platform:** Android

## Logs

```
 java.lang.RuntimeException: Unable to start activity ComponentInfo{com.amolg.flutter_app/com.amolg.flutter_app.MainActivity}: java.lang.ClassCastException: com.amolg.flutter_app.MainActivity cannot be cast to io.flutter.app.FlutterActivity
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2861)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2943)
        at android.app.ActivityThread.-wrap11(Unknown Source:0)
```

```
Analyzing flutter_app...                                                
No issues found! (ran in 17.5s)
```

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, v1.12.13+hotfix.5, on Mac OS X 10.15.2 19C57,
    locale en-IN)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 11.3)
[✓] Android Studio (version 3.5)
[!] VS Code (version 1.40.1)
    ✗ Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter
[✓] Connected device (1 available)

! Doctor found issues in 1 category.

```

**Comment by iapicca** [comment]
Hi @AmolGangadhare 
could you please run 
`flutter clean`
`flutter pub cache repair`
If the issue persist please provide your  `flutter doctor -v` and `flutter run --verbose`
together with a a minimal code sample to reproduce the problem
thank you

**Comment by AmolGangadhare** [comment]
@iapicca Thanks for your reply.

Tried `flutter clean` and `flutter pub cache repair` but it still not working.

1. `flutter doctor -v`
```[✓] Flutter (Channel stable, v1.12.13+hotfix.5, on Mac OS X 10.15.2 19C57, locale en-IN)
    • Flutter version 1.12.13+hotfix.5 at /Volumes/MyData/Flutter_SDK/flutter
    • Framework revision 27321ebbad (6 days ago), 2019-12-10 18:15:01 -0800
    • Engine revision 2994f7e1e6
    • Dart version 2.7.0

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/admin/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3, Build version 11C29
    • CocoaPods version 1.8.4

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 42.1.1
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[!] VS Code (version 1.40.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    ✗ Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (1 available)
    • Moto G 5 Plus • ZY22433CCB • android-arm • Android 8.1.0 (API 27)

! Doctor found issues in 1 category.
```

2. `flutter run --verbose` As it is too long so I'm adding the log of only after installing and before crashing. Do let me know if you need a complete log.

```
[   +9 ms] Installing build/app/outputs/apk/app.apk...
[        ] executing: /Users/admin/Library/Android/sdk/platform-tools/adb -s ZY22433CCB install -t -r
/Users/admin/Desktop/testFlutterApp/flutter_app/build/app/outputs/apk/app.apk
[+8036 ms] Performing Streamed Install
                    Success
[        ] Installing build/app/outputs/apk/app.apk... (completed in 8.0s)
[   +2 ms] executing: /Users/admin/Library/Android/sdk/platform-tools/adb -s ZY22433CCB shell echo -n d9e067103a99cd90331f475f4b153da5ca011041 >
/data/local/tmp/sky.com.amolg.flutter_app.sha1
[  +47 ms] Moto G 5 Plus startApp
[   +4 ms] executing: /Users/admin/Library/Android/sdk/platform-tools/adb -s ZY22433CCB shell am start -a android.intent.action.RUN -f 0x20000000
--ez enable-background-compilation true --ez enable-dart-profiling true --ez enable-checked-mode true --ez verify-entry-points true
com.amolg.flutter_app/com.amolg.flutter_app.MainActivity
[ +135 ms] Starting: Intent { act=android.intent.action.RUN flg=0x20000000 cmp=com.amolg.flutter_app/.MainActivity (has extras) }
[        ] Waiting for observatory port to be available...
[+1439 ms] Observatory URL on device: http://127.0.0.1:44491/5DHNFRb-YFs=/
[   +2 ms] executing: /Users/admin/Library/Android/sdk/platform-tools/adb -s ZY22433CCB forward tcp:0 tcp:44491
[  +11 ms] 57688
[        ] Forwarded host port 57688 to device port 44491 for Observatory
[  +14 ms] Connecting to service protocol: http://127.0.0.1:57688/5DHNFRb-YFs=/
[ +434 ms] Successfully connected to service protocol: http://127.0.0.1:57688/5DHNFRb-YFs=/
[   +5 ms] Sending to VM service: getVM({})
[  +10 ms] Result: {type: VM, name: vm, architectureBits: 32, hostCPU: Qualcomm Technologies, Inc MSM8953, operatingSystem: android, targetCPU: arm,
version: 2.7.0 (Mon Dec 2 20:10:59 2019 +0100) on "android_arm", _profilerMode: VM, _nativeZoneMemoryUsage: 0, pid: 7...
[   +8 ms] Sending to VM service: getIsolate({isolateId: isolates/2892632648324167})
[   +5 ms] Sending to VM service: _flutter.listViews({})
[  +23 ms] Result: {type: Isolate, id: isolates/2892632648324167, name: main, number: 2892632648324167, _originNumber: 2892632648324167, startTime:
1576572690640, _heaps: {new: {type: HeapSpace, name: new, vmName: Scavenger, collections: 1, avgCollectionPeriodMillis...
[   +6 ms] Result: {type: FlutterViewList, views: [{type: FlutterView, id: _flutterView/0xa5f35390, isolate: {type: @Isolate, fixedId: true, id:
isolates/2892632648324167, name: main.dart$main-2892632648324167, number: 2892632648324167}}]}
[   +6 ms] DevFS: Creating new filesystem on the device (null)
[   +1 ms] Sending to VM service: _createDevFS({fsName: flutter_app})
[  +72 ms] Result: {type: FileSystem, name: flutter_app, uri: file:///data/user/0/com.amolg.flutter_app/code_cache/flutter_appSDMGOQ/flutter_app/}
[        ] DevFS: Created new filesystem on the device (file:///data/user/0/com.amolg.flutter_app/code_cache/flutter_appSDMGOQ/flutter_app/)
[   +2 ms] Updating assets
[ +114 ms] Syncing files to device Moto G 5 Plus...
[   +3 ms] Scanning asset files
[   +3 ms] <- reset
[        ] Compiling dart to kernel with 0 updated files
[  +11 ms] /Volumes/MyData/Flutter_SDK/flutter/bin/cache/dart-sdk/bin/dart
/Volumes/MyData/Flutter_SDK/flutter/bin/cache/artifacts/engine/darwin-x64/frontend_server.dart.snapshot --sdk-root
/Volumes/MyData/Flutter_SDK/flutter/bin/cache/artifacts/engine/common/flutter_patched_sdk/ --incremental --target=flutter
-Ddart.developer.causal_async_stacks=true --output-dill /var/folders/xn/gn62hpdn3gjd7j4bp54452h40000gn/T/flutter_tool.qdKkY5/app.dill --packages
/Users/admin/Desktop/testFlutterApp/flutter_app/.packages -Ddart.vm.profile=false -Ddart.vm.product=false
--bytecode-options=source-positions,local-var-info,debugger-stops,instance-field-initializers,keep-unreachable-code,avoid-closure-call-instructions
--enable-asserts --track-widget-creation --filesystem-scheme org-dartlang-root
[   +5 ms] <- compile package:flutter_app/main.dart
[+7783 ms] Updating files
[ +197 ms] DevFS: Sync finished
[   +2 ms] Syncing files to device Moto G 5 Plus... (completed in 8,005ms, longer than expected)
[   +1 ms] Synced 0.9MB.
[   +2 ms] Sending to VM service: _flutter.listViews({})
[   +6 ms] Result: {type: FlutterViewList, views: [{type: FlutterView, id: _flutterView/0xa5f35390, isolate: {type: @Isolate, fixedId: true, id:
isolates/2892632648324167, name: main.dart$main-2892632648324167, number: 2892632648324167}}]}
[   +1 ms] <- accept
[        ] Connected to _flutterView/0xa5f35390.

```

3. Code

```
import io.flutter.app.FlutterActivity;
```
...

Crashing while casting to `FlutterActivity`
```
instance = new FlutterBarcodeScannerPlugin((FlutterActivity) registrar.activity(), registrar);
```

**Comment by xster** [comment]
My impression is that this issue is from the perspective of plugin development? If so, please see https://flutter.dev/go/android-plugin-migration for the new plugins API.

**Comment by AmolGangadhare** [comment]
Resolved by migrating to a new plugin API.

Thanks, @iapicca and @xster for your kind support.

**Comment by AmolGangadhare** [comment]
Surprisingly getting this error

```
E/flutter (17116): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: MissingPluginException(No implementation found for method scanBarcode on channel flutter_barcode_scanner)
```

Followed all the migration steps. It was working till yesterday and now getting this error.
I've tried all the possibilities. 

It starts occurring after I did `flutter upgrade` today

**Comment by matthew-carroll** [comment]
@xster based on the latest comment in the thread, this now sounds like it's probably a bug with a specific plugin. Does it sound like that to you, too? If so, should a new issue be opened with the appropriate description and labels?

**Comment by xster** [comment]
Can you open a new issue with the new error and paste all your console output from flutter run?

**Comment by vyapariapp** [comment]
https://stackoverflow.com/questions/60771003/lasscastexception-io-flutter-app-flutterpluginregistry-cannot-be-cast-to-io-flu
<BR>
<BR>
<BR>



java.lang.RuntimeException: Unable to create service io.flutter.plugins.firebasemessaging.FlutterFirebaseMessagingService: java.lang.ClassCastException: io.flutter.app.FlutterPluginRegistry cannot be cast to io.flutter.embedding.engine.FlutterEngine at android.app.ActivityThread.handleCreateService(ActivityThread.java:3976) at android.app.ActivityThread.-wrap5(Unknown Source:0) at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2109) at android.os.Handler.dispatchMessage(Handler.java:109) at android.os.Looper.loop(Looper.java:166) at android.app.ActivityThread.main(ActivityThread.java:7555) at java.lang.reflect.Method.invoke(Native Method) at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:469) at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:963) Caused by: java.lang.ClassCastException: io.flutter.app.FlutterPluginRegistry cannot be cast to io.flutter.embedding.engine.FlutterEngine at com.midhilaj.aquarium_shop.Application.registerWith(Application.java:25) at io.flutter.plugins.firebasemessaging.FlutterFirebaseMessagingService.startBackgroundIsolate(FlutterFirebaseMessagingService.java:164) at io.flutter.plugins.firebasemessaging.FlutterFirebaseMessagingService.onCreate(FlutterFirebaseMessagingService.java:77) at android.app.ActivityThread.handleCreateService(ActivityThread.java:3956) at android.app.ActivityThread.-wrap5(Unknown Source:0)  at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2109)  at android.os.Handler.dispatchMessage(Handler.java:109)  at android.os.Looper.loop(Looper.java:166)  at android.app.ActivityThread.main(ActivityThread.java:7555)  at java.lang.reflect.Method.invoke(Native Method)  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:469)  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:963)

 




D/FlutterActivity( 3682): Using the launch theme as normal theme. D/FlutterActivityAndFragmentDelegate( 3682): Setting up FlutterEngine. D/FlutterActivityAndFragmentDelegate( 3682): No preferred FlutterEngine was provided. Creating a new FlutterEngine for this FlutterFragment. D/FlutterActivityAndFragmentDelegate( 3682): Attaching FlutterEngine to the Activity that owns this Fragment. D/FlutterView( 3682): Detaching from a FlutterEngine: null D/FlutterView( 3682): Not attached to an engine. Doing nothing. D/FlutterView( 3682): Attaching to a FlutterEngine: io.flutter.embedding.engine.FlutterEngine@54b88a0 D/FlutterActivityAndFragmentDelegate( 3682): Executing Dart entrypoint: main, and sending initial route: / Syncing files to device DUB LX1... I/flutter ( 3682): -----=--token:db37iEIOPw2r5Y9LpYpXz-H V/FA ( 3682): Inactivity, disconnecting from the service I/flutter ( 3682): e222banner1 => {actionPage: null, buttonLabel: null, buttonLabelAr: null, img: https://firebasestorage.googleapis.com/, subTitle: null, subTitleAr: null, title: Gold FIsh, titleAr: null} I/flutter ( 3682): getMyCart fn I/flutter ( 3682): e222banner2 => {actionPage: null, buttonLabel: null, buttonLabelAr: null, img: https://firebasestorage.googleapis.com/ , subTitle: null, subTitleAr: null, title: Gold FIsh, titleAr: null} I/flutter ( 3682): getMyCart user 6AiQ2pG8nsaskpcUbbyJM7eCPuR2 D/ConnectivityManager( 3682): requestNetwork and the calling app is: com.midhilaj.aquarium_shop W/DynamiteModule( 3682): Local module descriptor class for providerinstaller not found. I/DynamiteModule( 3682): Considering local module providerinstaller:0 and remote module providerinstaller:0 W/ProviderInstaller( 3682): Failed to load providerinstaller module: No acceptable module found. Local version is 0 and remote version is 0. D/HwFLClassLoader( 3682): USE_FEATURE_LIST had not init! W/zygote64( 3682): ClassLoaderContext classpath size mismatch for position 0. expected=0, found=1 (PCL[] | PCL[/system/framework/com.android.media.remotedisplay.jar*818632460]) V/NativeCrypto( 3682): Registering com/google/android/gms/org/conscrypt/NativeCrypto's 286 native methods... D/NetworkSecurityConfig( 3682): No Network Security Config specified, using platform default I/ProviderInstaller( 3682): Installed default security provider GmsCore_OpenSSL D/ConnectivityManager( 3682): requestNetwork and the calling app is: com.midhilaj.aquarium_shop I/zygote64( 3682): Do partial code cache collection, code=60KB, data=46KB I/zygote64( 3682): After code cache collection, code=60KB, data=46KB I/zygote64( 3682): Increasing code cache capacity to 256KB I/zygote64( 3682): Starting profile saver IsSaveProfileNow end. I/flutter ( 3682): snapshot is null D/AndroidRuntime( 3682): Shutting down VM E/AndroidRuntime( 3682): FATAL EXCEPTION: main E/AndroidRuntime( 3682): Process: com.midhilaj.aquarium_shop, PID: 3682 E/AndroidRuntime( 3682): java.lang.RuntimeException: Unable to create service io.flutter.plugins.firebasemessaging.FlutterFirebaseMessagingService: java.lang.ClassCastException: io.flutter.app.FlutterPluginRegistry cannot be cast to io.flutter.embedding.engine.FlutterEngine E/AndroidRuntime( 3682): at android.app.ActivityThread.handleCreateService(ActivityThread.java:3976) E/AndroidRuntime( 3682): at android.app.ActivityThread.-wrap5(Unknown Source:0) E/AndroidRuntime( 3682): at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2109) E/AndroidRuntime( 3682): at android.os.Handler.dispatchMessage(Handler.java:109) E/AndroidRuntime( 3682): at android.os.Looper.loop(Looper.java:166) E/AndroidRuntime( 3682): at android.app.ActivityThread.main(ActivityThread.java:7555) E/AndroidRuntime( 3682): at java.lang.reflect.Method.invoke(Native Method) E/AndroidRuntime( 3682): at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:469) E/AndroidRuntime( 3682): at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:963) E/AndroidRuntime( 3682): Caused by: java.lang.ClassCastException: io.flutter.app.FlutterPluginRegistry cannot be cast to io.flutter.embedding.engine.FlutterEngine E/AndroidRuntime( 3682): at com.midhilaj.aquarium_shop.Application.registerWith(Application.java:25) E/AndroidRuntime( 3682): at io.flutter.plugins.firebasemessaging.FlutterFirebaseMessagingService.startBackgroundIsolate(FlutterFirebaseMessagingService.java:164) E/AndroidRuntime( 3682): at io.flutter.plugins.firebasemessaging.FlutterFirebaseMessagingService.onCreate(FlutterFirebaseMessagingService.java:77) E/AndroidRuntime( 3682): at android.app.ActivityThread.handleCreateService(ActivityThread.java:3956) E/AndroidRuntime( 3682): ... 8 more I/Process ( 3682): Sending signal. PID: 3682 SIG: 9 Lost connection to device.

import io.flutter.app.FlutterApplication;
import io.flutter.embedding.engine.FlutterEngine;
import io.flutter.plugin.common.PluginRegistry;
import io.flutter.plugin.common.PluginRegistry.PluginRegistrantCallback;

import io.flutter.app.FlutterApplication;
import io.flutter.plugin.common.PluginRegistry;
import io.flutter.plugin.common.PluginRegistry.PluginRegistrantCallback;
import io.flutter.plugins.GeneratedPluginRegistrant;
import io.flutter.plugins.firebasemessaging.FlutterFirebaseMessagingService;

public class Application extends FlutterApplication implements PluginRegistrantCallback {
    FlutterEngine fe;

    @Override
    public void onCreate() {
        super.onCreate();
        FlutterFirebaseMessagingService.setPluginRegistrant(this);
    }

    @Override
    public void registerWith(PluginRegistry registry) {
        GeneratedPluginRegistrant.registerWith((FlutterEngine) registry);
    }
}
